fl4health/clients/flash_client.py

```diff
@@ -0,0 +1,139 @@
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+    ) -> None:
+        """
+        This client is used to perform client-side training associated with the FLASH method described in
+        https://proceedings.mlr.press/v202/panchal23a/panchal23a.pdf. FLASH is designed to handle statistical
+        heterogeneity and concept drift in federated learning through client-side early stopping and server-side
+        drift-aware adaptive optimization.
+
+        Args:
+            data_path (Path): Path to the data directory.
+            metrics (Sequence[Metric]): List of metrics to be used for evaluation.
+            device (torch.device): Device to be used for training.
+            loss_meter_type (LossMeterType, optional):
+                Type of loss meter to be used. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[ClientCheckpointModule], optional): Checkpointer module defining when and how to do
+                checkpointing during client-side training. No checkpointing is done if not provided. Defaults to None.
+            gamma (float, optional):
+                Threshold for early stopping based on the change in validation loss. Defaults to 0.04.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.gamma: Optional[float] = None
+
+    def train_by_epochs(
+        self, epochs: int, current_round: Optional[int] = None
+    ) -> Tuple[Dict[str, float], Dict[str, Scalar]]:
+        self.model.train()
+        local_step = 0
+        previous_loss = float("inf")
+        for local_epoch in range(epochs):
+            self.train_metric_manager.clear()
+            self.train_loss_meter.clear()
+            for input, target in self.train_loader:
+                if self.is_empty_batch(input):
+                    log(INFO, "Empty batch generated by data loader. Skipping step.")
+                    continue
+
+                input, target = self._move_input_data_to_device(input), target.to(self.device)
+                losses, preds = self.train_step(input, target)
+                self.train_loss_meter.update(losses)
+                self.train_metric_manager.update(preds, target)
+                self.update_after_step(local_step)
+                self.total_steps += 1
+                local_step += 1
+
+            metrics = self.train_metric_manager.compute()
+            loss_dict = self.train_loss_meter.compute().as_dict()
+            current_loss = loss_dict.get("backward", 0.0)
+
+            # Early stopping check
+            if self.gamma is not None and abs(previous_loss - current_loss) < self.gamma / (local_step + 1):
```

Comment:
Finally, it looks like we're using the training loss here. In the paper, they use the validation loss to determine whether to perform early stopping. You should be able to get this loss value simply by calling `validate`.

Reply:
done in the latest commit